<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.16"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>ClaraGenomicsAnalysis: claragenomics::Index Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">ClaraGenomicsAnalysis
   &#160;<span id="projectnumber">0.2.0</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.16 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>claragenomics</b></li><li class="navelem"><a class="el" href="classclaragenomics_1_1Index.html">Index</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="classclaragenomics_1_1Index-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">claragenomics::Index Class Reference<span class="mlabels"><span class="mlabel">abstract</span></span><div class="ingroups"><a class="el" href="group__cudamapper.html">CUDA mapper package</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p><a class="el" href="classclaragenomics_1_1Index.html" title="Index - manages mapping of (k,w)-kmer-representation and all its occurences.">Index</a> - manages mapping of (k,w)-kmer-representation and all its occurences.  
 <a href="classclaragenomics_1_1Index.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="index_8hpp_source.html">index.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structclaragenomics_1_1Index_1_1RepresentationToSketchElements.html">RepresentationToSketchElements</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structclaragenomics_1_1Index_1_1RepresentationToSketchElements.html" title="RepresentationToSketchElements - representation, pointer to section of data arrays with sketch elemen...">RepresentationToSketchElements</a> - representation, pointer to section of data arrays with sketch elements with that representation and a given read_id, and a pointer to section of data arrays with sketch elements with that representation and all read_ids.  <a href="structclaragenomics_1_1Index_1_1RepresentationToSketchElements.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:ae329c278afa88376ae2bc9d23b825902"><td class="memItemLeft" align="right" valign="top"><a id="ae329c278afa88376ae2bc9d23b825902"></a>
virtual&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classclaragenomics_1_1Index.html#ae329c278afa88376ae2bc9d23b825902">~Index</a> ()=default</td></tr>
<tr class="memdesc:ae329c278afa88376ae2bc9d23b825902"><td class="mdescLeft">&#160;</td><td class="mdescRight">Virtual destructor for <a class="el" href="classclaragenomics_1_1Index.html" title="Index - manages mapping of (k,w)-kmer-representation and all its occurences.">Index</a>. <br /></td></tr>
<tr class="separator:ae329c278afa88376ae2bc9d23b825902"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af5193b086fdbe74cb99c111e5db1d505"><td class="memItemLeft" align="right" valign="top">virtual const std::vector&lt; position_in_read_t &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classclaragenomics_1_1Index.html#af5193b086fdbe74cb99c111e5db1d505">positions_in_reads</a> () const =0</td></tr>
<tr class="memdesc:af5193b086fdbe74cb99c111e5db1d505"><td class="mdescLeft">&#160;</td><td class="mdescRight">returns an array of starting positions of sketch elements in their reads  <a href="classclaragenomics_1_1Index.html#af5193b086fdbe74cb99c111e5db1d505">More...</a><br /></td></tr>
<tr class="separator:af5193b086fdbe74cb99c111e5db1d505"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afd734a8401e11ed1a4e14dba5bb2c8ae"><td class="memItemLeft" align="right" valign="top">virtual const std::vector&lt; read_id_t &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classclaragenomics_1_1Index.html#afd734a8401e11ed1a4e14dba5bb2c8ae">read_ids</a> () const =0</td></tr>
<tr class="memdesc:afd734a8401e11ed1a4e14dba5bb2c8ae"><td class="mdescLeft">&#160;</td><td class="mdescRight">returns an array of reads ids for sketch elements  <a href="classclaragenomics_1_1Index.html#afd734a8401e11ed1a4e14dba5bb2c8ae">More...</a><br /></td></tr>
<tr class="separator:afd734a8401e11ed1a4e14dba5bb2c8ae"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0ff648f8c3a83e0dd838a160f00ba5bd"><td class="memItemLeft" align="right" valign="top">virtual const std::vector&lt; <a class="el" href="classclaragenomics_1_1SketchElement.html#a49c2500e86ac9fdb4d5bad6ad2b47f79">SketchElement::DirectionOfRepresentation</a> &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classclaragenomics_1_1Index.html#a0ff648f8c3a83e0dd838a160f00ba5bd">directions_of_reads</a> () const =0</td></tr>
<tr class="memdesc:a0ff648f8c3a83e0dd838a160f00ba5bd"><td class="mdescLeft">&#160;</td><td class="mdescRight">returns an array of directions in which sketch elements were read  <a href="classclaragenomics_1_1Index.html#a0ff648f8c3a83e0dd838a160f00ba5bd">More...</a><br /></td></tr>
<tr class="separator:a0ff648f8c3a83e0dd838a160f00ba5bd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab8f53484c09c375c64c1bd963719dbe5"><td class="memItemLeft" align="right" valign="top">virtual std::uint64_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classclaragenomics_1_1Index.html#ab8f53484c09c375c64c1bd963719dbe5">number_of_reads</a> () const =0</td></tr>
<tr class="memdesc:ab8f53484c09c375c64c1bd963719dbe5"><td class="mdescLeft">&#160;</td><td class="mdescRight">returns number of reads in input data  <a href="classclaragenomics_1_1Index.html#ab8f53484c09c375c64c1bd963719dbe5">More...</a><br /></td></tr>
<tr class="separator:ab8f53484c09c375c64c1bd963719dbe5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1ef291e4ea3a2e7bcb236785362813fa"><td class="memItemLeft" align="right" valign="top">virtual const std::vector&lt; std::string &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classclaragenomics_1_1Index.html#a1ef291e4ea3a2e7bcb236785362813fa">read_id_to_read_name</a> () const =0</td></tr>
<tr class="memdesc:a1ef291e4ea3a2e7bcb236785362813fa"><td class="mdescLeft">&#160;</td><td class="mdescRight">returns mapping of internal read id that goes from 0 to number_of_reads-1 to actual read name from the input  <a href="classclaragenomics_1_1Index.html#a1ef291e4ea3a2e7bcb236785362813fa">More...</a><br /></td></tr>
<tr class="separator:a1ef291e4ea3a2e7bcb236785362813fa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abb583d6038cd9e1fc39416ba3ea11f0f"><td class="memItemLeft" align="right" valign="top">virtual const std::vector&lt; std::uint32_t &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classclaragenomics_1_1Index.html#abb583d6038cd9e1fc39416ba3ea11f0f">read_id_to_read_length</a> () const =0</td></tr>
<tr class="memdesc:abb583d6038cd9e1fc39416ba3ea11f0f"><td class="mdescLeft">&#160;</td><td class="mdescRight">returns mapping of internal read id that goes from 0 to read lengths for that read  <a href="classclaragenomics_1_1Index.html#abb583d6038cd9e1fc39416ba3ea11f0f">More...</a><br /></td></tr>
<tr class="separator:abb583d6038cd9e1fc39416ba3ea11f0f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acc4e0f7ede68c0c46d47822f521d45cf"><td class="memItemLeft" align="right" valign="top">virtual std::uint64_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classclaragenomics_1_1Index.html#acc4e0f7ede68c0c46d47822f521d45cf">minimum_representation</a> () const =0</td></tr>
<tr class="memdesc:acc4e0f7ede68c0c46d47822f521d45cf"><td class="mdescLeft">&#160;</td><td class="mdescRight">minimum possible representation  <a href="classclaragenomics_1_1Index.html#acc4e0f7ede68c0c46d47822f521d45cf">More...</a><br /></td></tr>
<tr class="separator:acc4e0f7ede68c0c46d47822f521d45cf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afe36fc8a01ae01b3619b10f577a83daf"><td class="memItemLeft" align="right" valign="top">virtual std::uint64_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classclaragenomics_1_1Index.html#afe36fc8a01ae01b3619b10f577a83daf">maximum_representation</a> () const =0</td></tr>
<tr class="memdesc:afe36fc8a01ae01b3619b10f577a83daf"><td class="mdescLeft">&#160;</td><td class="mdescRight">maximum possible representation  <a href="classclaragenomics_1_1Index.html#afe36fc8a01ae01b3619b10f577a83daf">More...</a><br /></td></tr>
<tr class="separator:afe36fc8a01ae01b3619b10f577a83daf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af7094b57c159411265364b36e4b49c61"><td class="memItemLeft" align="right" valign="top">virtual const std::vector&lt; std::vector&lt; <a class="el" href="structclaragenomics_1_1Index_1_1RepresentationToSketchElements.html">RepresentationToSketchElements</a> &gt; &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classclaragenomics_1_1Index.html#af7094b57c159411265364b36e4b49c61">read_id_and_representation_to_sketch_elements</a> () const =0</td></tr>
<tr class="memdesc:af7094b57c159411265364b36e4b49c61"><td class="mdescLeft">&#160;</td><td class="mdescRight">For each read_id (outer vector) returns a vector in which each element contains a representation from that read, pointer to section of data arrays with sketch elements with that representation and that read_id, and pointer to section of data arrays with skecth elements with that representation and all read_ids. There elements are sorted by representation in increasing order.  <a href="classclaragenomics_1_1Index.html#af7094b57c159411265364b36e4b49c61">More...</a><br /></td></tr>
<tr class="separator:af7094b57c159411265364b36e4b49c61"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4bc6d086fc48947e137fe0dcae9ddf51"><td class="memItemLeft" align="right" valign="top">virtual bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classclaragenomics_1_1Index.html#a4bc6d086fc48947e137fe0dcae9ddf51">reached_end_of_input</a> () const =0</td></tr>
<tr class="memdesc:a4bc6d086fc48947e137fe0dcae9ddf51"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns whether there are any more reads to process in the reads file (e.g FASTA file)  <a href="classclaragenomics_1_1Index.html#a4bc6d086fc48947e137fe0dcae9ddf51">More...</a><br /></td></tr>
<tr class="separator:a4bc6d086fc48947e137fe0dcae9ddf51"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:a68e67da53783a078d9352fe586fa3433"><td class="memItemLeft" align="right" valign="top">static std::unique_ptr&lt; <a class="el" href="classclaragenomics_1_1Index.html">Index</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classclaragenomics_1_1Index.html#a68e67da53783a078d9352fe586fa3433">create_index</a> (const std::string &amp;query_filename, const std::uint64_t kmer_size, const std::uint64_t window_size, const std::vector&lt; std::pair&lt; std::uint64_t, std::uint64_t &gt;&gt; &amp;ranges)</td></tr>
<tr class="memdesc:a68e67da53783a078d9352fe586fa3433"><td class="mdescLeft">&#160;</td><td class="mdescRight">generates a mapping of (k,w)-kmer-representation to all of its occurrences for one or more sequences  <a href="classclaragenomics_1_1Index.html#a68e67da53783a078d9352fe586fa3433">More...</a><br /></td></tr>
<tr class="separator:a68e67da53783a078d9352fe586fa3433"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adf493e4378c6c01323e838a1dcab2fb9"><td class="memItemLeft" align="right" valign="top">static std::unique_ptr&lt; <a class="el" href="classclaragenomics_1_1Index.html">Index</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classclaragenomics_1_1Index.html#adf493e4378c6c01323e838a1dcab2fb9">create_index</a> ()</td></tr>
<tr class="memdesc:adf493e4378c6c01323e838a1dcab2fb9"><td class="mdescLeft">&#160;</td><td class="mdescRight">creates an empty <a class="el" href="classclaragenomics_1_1Index.html" title="Index - manages mapping of (k,w)-kmer-representation and all its occurences.">Index</a>  <a href="classclaragenomics_1_1Index.html#adf493e4378c6c01323e838a1dcab2fb9">More...</a><br /></td></tr>
<tr class="separator:adf493e4378c6c01323e838a1dcab2fb9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a570a64375c356f8c6d745c55bf90bf5a"><td class="memItemLeft" align="right" valign="top">static uint64_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classclaragenomics_1_1Index.html#a570a64375c356f8c6d745c55bf90bf5a">maximum_kmer_size</a> ()</td></tr>
<tr class="memdesc:a570a64375c356f8c6d745c55bf90bf5a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the maximum kmer length allowable. This is just the size of the representation in bits divided by two (since 2 bits are required to represent a DNA nucleotide).  <a href="classclaragenomics_1_1Index.html#a570a64375c356f8c6d745c55bf90bf5a">More...</a><br /></td></tr>
<tr class="separator:a570a64375c356f8c6d745c55bf90bf5a"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p><a class="el" href="classclaragenomics_1_1Index.html" title="Index - manages mapping of (k,w)-kmer-representation and all its occurences.">Index</a> - manages mapping of (k,w)-kmer-representation and all its occurences. </p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="adf493e4378c6c01323e838a1dcab2fb9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#adf493e4378c6c01323e838a1dcab2fb9">&#9670;&nbsp;</a></span>create_index() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static std::unique_ptr&lt;<a class="el" href="classclaragenomics_1_1Index.html">Index</a>&gt; claragenomics::Index::create_index </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>creates an empty <a class="el" href="classclaragenomics_1_1Index.html" title="Index - manages mapping of (k,w)-kmer-representation and all its occurences.">Index</a> </p>
<dl class="section return"><dt>Returns</dt><dd>empty instacne of <a class="el" href="classclaragenomics_1_1Index.html" title="Index - manages mapping of (k,w)-kmer-representation and all its occurences.">Index</a> </dd></dl>

</div>
</div>
<a id="a68e67da53783a078d9352fe586fa3433"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a68e67da53783a078d9352fe586fa3433">&#9670;&nbsp;</a></span>create_index() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static std::unique_ptr&lt;<a class="el" href="classclaragenomics_1_1Index.html">Index</a>&gt; claragenomics::Index::create_index </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>query_filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::uint64_t&#160;</td>
          <td class="paramname"><em>kmer_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::uint64_t&#160;</td>
          <td class="paramname"><em>window_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::vector&lt; std::pair&lt; std::uint64_t, std::uint64_t &gt;&gt; &amp;&#160;</td>
          <td class="paramname"><em>ranges</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>generates a mapping of (k,w)-kmer-representation to all of its occurrences for one or more sequences </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">query_filename</td><td>filepath to reads in FASTA or FASTQ format </td></tr>
    <tr><td class="paramname">kmer_size</td><td>k - the kmer length </td></tr>
    <tr><td class="paramname">window_size</td><td>w - the length of the sliding window used to find sketch elements </td></tr>
    <tr><td class="paramname">ranges</td><td>- the ranges of reads in the query file to use for mapping, index by their position (e.g in the FASTA file) </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>instance of <a class="el" href="classclaragenomics_1_1Index.html" title="Index - manages mapping of (k,w)-kmer-representation and all its occurences.">Index</a> </dd></dl>

</div>
</div>
<a id="a0ff648f8c3a83e0dd838a160f00ba5bd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0ff648f8c3a83e0dd838a160f00ba5bd">&#9670;&nbsp;</a></span>directions_of_reads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">virtual const std::vector&lt;<a class="el" href="classclaragenomics_1_1SketchElement.html#a49c2500e86ac9fdb4d5bad6ad2b47f79">SketchElement::DirectionOfRepresentation</a>&gt;&amp; claragenomics::Index::directions_of_reads </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">pure virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>returns an array of directions in which sketch elements were read </p>
<dl class="section return"><dt>Returns</dt><dd>an array of directions in which sketch elements were read </dd></dl>

</div>
</div>
<a id="a570a64375c356f8c6d745c55bf90bf5a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a570a64375c356f8c6d745c55bf90bf5a">&#9670;&nbsp;</a></span>maximum_kmer_size()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static uint64_t claragenomics::Index::maximum_kmer_size </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return the maximum kmer length allowable. This is just the size of the representation in bits divided by two (since 2 bits are required to represent a DNA nucleotide). </p>
<dl class="section return"><dt>Returns</dt><dd>Return the maximum kmer length allowable </dd></dl>

</div>
</div>
<a id="afe36fc8a01ae01b3619b10f577a83daf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afe36fc8a01ae01b3619b10f577a83daf">&#9670;&nbsp;</a></span>maximum_representation()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">virtual std::uint64_t claragenomics::Index::maximum_representation </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">pure virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>maximum possible representation </p>
<dl class="section return"><dt>Returns</dt><dd>the largest possible representation </dd></dl>

</div>
</div>
<a id="acc4e0f7ede68c0c46d47822f521d45cf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acc4e0f7ede68c0c46d47822f521d45cf">&#9670;&nbsp;</a></span>minimum_representation()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">virtual std::uint64_t claragenomics::Index::minimum_representation </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">pure virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>minimum possible representation </p>
<dl class="section return"><dt>Returns</dt><dd>the smallest possible representation </dd></dl>

</div>
</div>
<a id="ab8f53484c09c375c64c1bd963719dbe5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab8f53484c09c375c64c1bd963719dbe5">&#9670;&nbsp;</a></span>number_of_reads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">virtual std::uint64_t claragenomics::Index::number_of_reads </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">pure virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>returns number of reads in input data </p>
<dl class="section return"><dt>Returns</dt><dd>number of reads in input data </dd></dl>

</div>
</div>
<a id="af5193b086fdbe74cb99c111e5db1d505"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af5193b086fdbe74cb99c111e5db1d505">&#9670;&nbsp;</a></span>positions_in_reads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">virtual const std::vector&lt;position_in_read_t&gt;&amp; claragenomics::Index::positions_in_reads </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">pure virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>returns an array of starting positions of sketch elements in their reads </p>
<dl class="section return"><dt>Returns</dt><dd>an array of starting positions of sketch elements in their reads </dd></dl>

</div>
</div>
<a id="a4bc6d086fc48947e137fe0dcae9ddf51"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4bc6d086fc48947e137fe0dcae9ddf51">&#9670;&nbsp;</a></span>reached_end_of_input()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">virtual bool claragenomics::Index::reached_end_of_input </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">pure virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Returns whether there are any more reads to process in the reads file (e.g FASTA file) </p>
<dl class="section return"><dt>Returns</dt><dd>Returns whether there are any more reads to process in the reads file (e.g FASTA file) </dd></dl>

</div>
</div>
<a id="af7094b57c159411265364b36e4b49c61"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af7094b57c159411265364b36e4b49c61">&#9670;&nbsp;</a></span>read_id_and_representation_to_sketch_elements()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">virtual const std::vector&lt;std::vector&lt;<a class="el" href="structclaragenomics_1_1Index_1_1RepresentationToSketchElements.html">RepresentationToSketchElements</a>&gt; &gt;&amp; claragenomics::Index::read_id_and_representation_to_sketch_elements </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">pure virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>For each read_id (outer vector) returns a vector in which each element contains a representation from that read, pointer to section of data arrays with sketch elements with that representation and that read_id, and pointer to section of data arrays with skecth elements with that representation and all read_ids. There elements are sorted by representation in increasing order. </p>
<dl class="section return"><dt>Returns</dt><dd>the mapping </dd></dl>

</div>
</div>
<a id="abb583d6038cd9e1fc39416ba3ea11f0f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abb583d6038cd9e1fc39416ba3ea11f0f">&#9670;&nbsp;</a></span>read_id_to_read_length()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">virtual const std::vector&lt;std::uint32_t&gt;&amp; claragenomics::Index::read_id_to_read_length </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">pure virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>returns mapping of internal read id that goes from 0 to read lengths for that read </p>
<dl class="section return"><dt>Returns</dt><dd>mapping of internal read id that goes from 0 to read lengths for that read </dd></dl>

</div>
</div>
<a id="a1ef291e4ea3a2e7bcb236785362813fa"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1ef291e4ea3a2e7bcb236785362813fa">&#9670;&nbsp;</a></span>read_id_to_read_name()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">virtual const std::vector&lt;std::string&gt;&amp; claragenomics::Index::read_id_to_read_name </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">pure virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>returns mapping of internal read id that goes from 0 to number_of_reads-1 to actual read name from the input </p>
<dl class="section return"><dt>Returns</dt><dd>mapping of internal read id that goes from 0 to number_of_reads-1 to actual read name from the input </dd></dl>

</div>
</div>
<a id="afd734a8401e11ed1a4e14dba5bb2c8ae"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afd734a8401e11ed1a4e14dba5bb2c8ae">&#9670;&nbsp;</a></span>read_ids()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">virtual const std::vector&lt;read_id_t&gt;&amp; claragenomics::Index::read_ids </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">pure virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>returns an array of reads ids for sketch elements </p>
<dl class="section return"><dt>Returns</dt><dd>an array of reads ids for sketch elements </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>cudamapper/include/cudamapper/<a class="el" href="index_8hpp_source.html">index.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.16
</small></address>
</body>
</html>
